src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -3424,52 +3424,445 @@ bool TheoryStringsRewriter::checkEntailArith(Node a, Node b, bool strict)
+  aar = Rewriter::rewrite( aar );
+  Trace("strings-ent-approx-debug") << "...processed fixed sum " << aar << " with " << mApprox.size() << " approximated monomials." << std::endl;
+  // if we have a choice of how to approximate
+  if( !mApprox.empty() )
+  {
+    // convert finaln back to monomial sum
+    std::map< Node, Node > msumAar;
+    if( !ArithMSum::getMonomialSum(aar,msumAar) )
+    {        
+      return false;
+    }
+    if( Trace.isOn("strings-ent-approx") )
+    {
+      Trace("strings-ent-approx") << "---- Check arithmetic entailment by under-approximation " << ar << " >= 0" << std::endl;
+      Trace("strings-ent-approx") << "FIXED:" << std::endl;
+      ArithMSum::debugPrintMonomialSum( msumAar, "strings-ent-approx" );
+      Trace("strings-ent-approx") << "APPROX:" << std::endl;
+      for( std::pair< const Node, std::vector< Node > >& a : mApprox )
+      {
+        Node c = msum[a.first];
+        Trace("strings-ent-approx") << "  ";
+        if( !c.isNull() )
+        {
+          Trace("strings-ent-approx") << c << " * ";
+        }
+        Trace("strings-ent-approx") << a.second << " ...from " << a.first << std::endl;
+      }
+      Trace("strings-ent-approx") << std::endl;
+    }
+    Rational one(1);
+    // incorporate monomials one at a time that have a choice of approximations
+    while( !mApprox.empty() )
+    {
+      Node v;
+      Node vapprox;
+      int maxScore = -1;
+      // look at each approximation, take the one with the best score
+      for( std::pair< const Node, std::vector< Node > >& nam : mApprox )
+      {
+        for( const Node& aa : nam.second )
+        {
+          unsigned helpsCancelCount = 0;
+          unsigned addsObligationCount = 0;
+          std::map< Node, Node >::iterator it;
+          for( std::pair< const Node, Node >& aam : approxMsums[aa] )
+          {
+            // Say aar is of the form t + c1*v, and aam is the monomial c2*v 
+            // where c2 != 0. We say aam:
+            // (1) helps cancel if c1 != 0 and c1>0 != c2>0
+            // (2) adds obligation if c1>=0 and c1+c2<0
+            Node v = aam.first;
+            Node c2 = aam.second;
+            int c2Sgn = c2.isNull() ? 1 : c2.getConst<Rational>().sgn();
+            it = msumAar.find(v);
+            if( it!=msumAar.end() )
+            {
+              Node c1 = it->second;
+              int c1Sgn = c1.isNull() ? 1 : c1.getConst<Rational>().sgn();
+              if( c1Sgn==0 )
+              {
+                addsObligationCount += ( c2Sgn==-1  ? 1 : 0 );
+              }
+              else if( c1Sgn!=c2Sgn )
+              {
+                helpsCancelCount++;
+                Rational r1 = c1.isNull() ? one : c1.getConst<Rational>();
+                Rational r2 = c2.isNull() ? one : c2.getConst<Rational>();
+                Rational r12 = r1 + r2;
+                if( r12.sgn()==-1 )
+                {
+                  addsObligationCount++;
```

Comment:
Hmm, so `aam` can both increase the `helpsCancelCount` and the `addsObligationCount` simultaneously?

Reply:
Yes. The idea is that for x+2*y, we consider in descending order of preference: -2*y,  -2*x, -z. The first cancels, the second cancels and adds obligation, the third only adds obligation.